### **Importing Libraries**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
import sqlite3
from scipy.stats import ttest_ind
import scipy.stats as stats
warnings.filterwarnings('ignore')

In [2]:
#creating databse connection

DB_PATH = r"C:\Users\lenovo\inventory.db"
conn = sqlite3.connect(DB_PATH)

print("Connected to:", DB_PATH)

# fetching vendor summary data
df = pd.read_sql_query(
    "SELECT * FROM vendor_sales_summary",
    conn
)

df.head()

Connected to: C:\Users\lenovo\inventory.db


VendorNumber                VendorName  Brand              Description  \
0          1128         BROWN-FORMAN CORP   1233  Jack Daniels No 7 Black   
1          4425     MARTIGNETTI COMPANIES   3405    Tito's Handmade Vodka   
2         17035         PERNOD RICARD USA   8068         Absolut 80 Proof   
3          3960  DIAGEO NORTH AMERICA INC   4261   Capt Morgan Spiced Rum   
4          3960  DIAGEO NORTH AMERICA INC   3545          Ketel One Vodka   

   PurchasePrice  ActualPrice  Volume  TotalPurchaseQuantity  \
0          26.27        36.99  1750.0                 145080   
1          23.19        28.99  1750.0                 164038   
2          18.24        24.99  1750.0                 187407   
3          16.17        22.99  1750.0                 201682   
4          21.89        29.99  1750.0                 138109   

   TotalPurchaseDollars  TotalSalesQuantity  TotalSalesDollars  \
0            3811251.60            142049.0         5101919.51   
1            3804041.22            160247.0         4819073.49   
2            3418303.68            187140.0         4538120.60   
3            3261197.94            200412.0         4475972.88   
4            3023206.01            135838.0         4223107.62   

   TotalSalesPrice  TotalExciseTax  FreightCost  GrossProfit  ProfitMargin  \
0        672819.31       260999.20     68601.68   1290667.91     25.297693   
1        561512.37       294438.66    144929.24   1015032.27     21.062810   
2        461140.15       343854.07    123780.22   1119816.92     24.675786   
3        420050.01       368242.80    257032.07   1214774.94     27.139908   
4        545778.28       249587.83    257032.07   1199901.61     28.412764   

   StockTurnover  SalesToPurchaseRatio  
0       0.979108              1.338647  
1       0.976890              1.266830  
2       0.998575              1.327594  
3       0.993703              1.372493  
4       0.983556              1.396897

### **Exploratory Data Analysis** 

- Previously, we examined the various tables in the database to identify key variables, understand their relationships, and determine which ones should be included in the final analysis.
- In this phase of EDA, we will analyze the resultant table to gain insights into the distribution of each column. This will help us understand data patterns, identify anomalies, and ensure data quality before proceeding with further analysis. 

In [3]:
#summary statistics
df.describe().T

count          mean            std           min  \
VendorNumber           10692.0  10650.649458   18753.519148      2.000000   
Brand                  10692.0  18039.228769   12662.187074     58.000000   
PurchasePrice          10692.0     24.385303     109.269375      0.360000   
ActualPrice            10692.0     35.643671     148.246016      0.490000   
Volume                 10692.0    847.360550     664.309212     50.000000   
TotalPurchaseQuantity  10692.0   3140.886831   11095.086769      1.000000   
TotalPurchaseDollars   10692.0  30106.693372  123067.799627      0.710000   
TotalSalesQuantity     10692.0   3077.482136   10952.851391      0.000000   
TotalSalesDollars      10692.0  42239.074419  167655.265984      0.000000   
TotalSalesPrice        10692.0  18793.783627   44952.773386      0.000000   
TotalExciseTax         10692.0   1774.226259   10975.582240      0.000000   
FreightCost            10692.0  61433.763214   60938.458032      0.090000   
GrossProfit            10692.0  12132.381048   46224.337964 -52002.780000   
ProfitMargin           10692.0    -15.620770     443.555329 -23730.638953   
StockTurnover          10692.0      1.706793       6.020460      0.000000   
SalesToPurchaseRatio   10692.0      2.504390       8.459067      0.000000   

                                25%           50%           75%           max  
VendorNumber            3951.000000   7153.000000   9552.000000  2.013590e+05  
Brand                   5793.500000  18761.500000  25514.250000  9.063100e+04  
PurchasePrice              6.840000     10.455000     19.482500  5.681810e+03  
ActualPrice               10.990000     15.990000     28.990000  7.499990e+03  
Volume                   750.000000    750.000000    750.000000  2.000000e+04  
TotalPurchaseQuantity     36.000000    262.000000   1975.750000  3.376600e+05  
TotalPurchaseDollars     453.457500   3655.465000  20738.245000  3.811252e+06  
TotalSalesQuantity        33.000000    261.000000   1929.250000  3.349390e+05  
TotalSalesDollars        729.220000   5298.045000  28396.915000  5.101920e+06  
TotalSalesPrice          289.710000   2857.800000  16059.562500  6.728193e+05  
TotalExciseTax             4.800000     46.570000    418.650000  3.682428e+05  
FreightCost            14069.870000  50293.620000  79528.990000  2.570321e+05  
GrossProfit               52.920000   1399.640000   8660.200000  1.290668e+06  
ProfitMargin              13.324515     30.405457     39.956135  9.971666e+01  
StockTurnover              0.807229      0.981529      1.039342  2.745000e+02  
SalesToPurchaseRatio       1.153729      1.436894      1.665449  3.529286e+02

In [4]:
import os

print("Current working directory:", os.getcwd())
print("Files/folders here:", os.listdir())

os.makedirs("outputs", exist_ok=True)

print("After creating outputs:", os.listdir())

Current working directory: C:\Users\lenovo
Files/folders here: ['.ipynb_checkpoints', '.ipython', '.jupyter', '.matplotlib', '3D Objects', 'AppData', 'Application Data', 'Contacts', 'Cookies', 'data', 'Documents', 'Downloads', 'Exploratory Data Analysis.ipynb', 'Favorites', 'get_vendor_summary.py', 'ingestion_db.py', 'IntelGraphicsProfiles', 'inventory.db', 'Links', 'Local Settings', 'Logs', 'Music', 'My Documents', 'NetHood', 'NTUSER.DAT', 'ntuser.dat.LOG1', 'ntuser.dat.LOG2', 'NTUSER.DAT{53b39e88-18c4-11ea-a811-000d3aa4692b}.TM.blf', 'NTUSER.DAT{53b39e88-18c4-11ea-a811-000d3aa4692b}.TMContainer00000000000000000001.regtrans-ms', 'NTUSER.DAT{53b39e88-18c4-11ea-a811-000d3aa4692b}.TMContainer00000000000000000002.regtrans-ms', 'ntuser.ini', 'OneDrive', 'outputs', 'PrintHood', 'Recent', 'Saved Games', 'Searches', 'SendTo', 'Start Menu', 'Templates', 'Vendor Performance Analysis.ipynb', 'vendor_sales_summary_clean.csv', 'Videos', '__pycache__']
After creating outputs: ['.ipynb_checkpoints',

In [5]:
# Distribution plots for Numerical Columns (saved separately)

save_dir = r"outputs\Histplot"
os.makedirs(save_dir, exist_ok=True)

numerical_cols = df.select_dtypes(include=np.number).columns

for col in numerical_cols:
    plt.figure(figsize=(6,4))
    sns.histplot(df[col], kde=True, bins=30)
    plt.title(f"Distribution of {col}")

    plt.savefig(os.path.join(save_dir, f"{col}_distribution.png"), dpi=300, bbox_inches="tight")
    plt.close()

In [6]:
#outlierdetection with BOXplots
 

# pick categorical columns automatically (object/category)
categorical_cols = df.select_dtypes(include=["object", "category"]).columns

save_dir = r"outputs\countplot"
os.makedirs(save_dir, exist_ok=True)

for col in categorical_cols:
    # if a column has too many unique values, top 10 makes sense
    top_n = 10
    order = df[col].value_counts(dropna=False).head(top_n).index

    plt.figure(figsize=(12, 6))
    sns.countplot(y=df[col], order=order)
    plt.title(f"Top {top_n} Count Plot of {col}")
    plt.tight_layout()

    plt.savefig(os.path.join(save_dir, f"{col}_countplot.png"), dpi=300, bbox_inches="tight")
    plt.close()   # closes the figure so notebook stays light

## **Summary Statistics Insights**

### **Negative & Zero Values:**
- **Gross Profit**: Minimum value is -52,002.78, indicating losses. Some products or transactions may be selling at a loss due to high costs or selling at discounts lower than the purchase price.
- **Profit Margin**: Has a minimum of -∞, which suggests cases where revenue is zero or even lower than costs.
- **Total Sales Quantity & Sales Dollars**: Minimum values are 0, meaning some products were purchased but never sold. These could be slow-moving or obsolete stock.

### **Outliers Indicated by High Standard Deviations:**

- **Purchase & Actual Prices**: The maximum values (5,681.81 & 7,499.99) are significantly higher than the mean (24.39 & 35.64), indicating potential premium products.
- **Freight Cost**: Huge variation, from 0.09 to 257,032.07, suggests logistics inefficiencies or bulk shipments.
- **Stock Turnover**: Ranges from 0 to 274.5, implying some products sell extremely fast while others remain in stock indefinitely. A value more than 1 indicates that the sold quantity for that product is higher than the purchased quantity, possibly due to sales being fulfilled from older stock.

In [7]:
# let's filter the data by removing inconsistencies
df = pd.read_sql_query("""SELECT * 
FROM vendor_sales_summary
WHERE GrossProfit > 0
AND ProfitMargin > 0 
AND TotalSalesQuantity > 0""", conn)

In [8]:
df

VendorNumber                VendorName  Brand  \
0             1128         BROWN-FORMAN CORP   1233   
1             4425     MARTIGNETTI COMPANIES   3405   
2            17035         PERNOD RICARD USA   8068   
3             3960  DIAGEO NORTH AMERICA INC   4261   
4             3960  DIAGEO NORTH AMERICA INC   3545   
...            ...                       ...    ...   
8559          9815            WINE GROUP INC   8527   
8560          8004            SAZERAC CO INC   5683   
8561          3924  HEAVEN HILL DISTILLERIES   9123   
8562          3960  DIAGEO NORTH AMERICA INC   6127   
8563          7245      PROXIMO SPIRITS INC.   3065   

                      Description  PurchasePrice  ActualPrice  Volume  \
0         Jack Daniels No 7 Black          26.27        36.99  1750.0   
1           Tito's Handmade Vodka          23.19        28.99  1750.0   
2                Absolut 80 Proof          18.24        24.99  1750.0   
3          Capt Morgan Spiced Rum          16.17        22.99  1750.0   
4                 Ketel One Vodka          21.89        29.99  1750.0   
...                           ...            ...          ...     ...   
8559  Concannon Glen Ellen Wh Zin           1.32         4.99   750.0   
8560  Dr McGillicuddy's Apple Pie           0.39         0.49    50.0   
8561              Deep Eddy Vodka           0.74         0.99    50.0   
8562  The Club Strawbry Margarita           1.47         1.99   200.0   
8563     Three Olives Grape Vodka           0.71         0.99    50.0   

      TotalPurchaseQuantity  TotalPurchaseDollars  TotalSalesQuantity  \
0                    145080            3811251.60            142049.0   
1                    164038            3804041.22            160247.0   
2                    187407            3418303.68            187140.0   
3                    201682            3261197.94            200412.0   
4                    138109            3023206.01            135838.0   
...                     ...                   ...                 ...   
8559                      2                  2.64                 5.0   
8560                      6                  2.34               134.0   
8561                      2                  1.48                 2.0   
8562                      1                  1.47                72.0   
8563                      1                  0.71                86.0   

      TotalSalesDollars  TotalSalesPrice  TotalExciseTax  FreightCost  \
0            5101919.51        672819.31       260999.20     68601.68   
1            4819073.49        561512.37       294438.66    144929.24   
2            4538120.60        461140.15       343854.07    123780.22   
3            4475972.88        420050.01       368242.80    257032.07   
4            4223107.62        545778.28       249587.83    257032.07   
...                 ...              ...             ...          ...   
8559              15.95            10.96            0.55     27100.41   
8560              65.66             1.47            7.04     50293.62   
8561               1.98             0.99            0.10     14069.87   
8562             143.28            77.61           15.12    257032.07   
8563              85.14            33.66            4.46     38994.78   

      GrossProfit  ProfitMargin  StockTurnover  SalesToPurchaseRatio  
0      1290667.91     25.297693       0.979108              1.338647  
1      1015032.27     21.062810       0.976890              1.266830  
2      1119816.92     24.675786       0.998575              1.327594  
3      1214774.94     27.139908       0.993703              1.372493  
4      1199901.61     28.412764       0.983556              1.396897  
...           ...           ...            ...                   ...  
8559        13.31     83.448276       2.500000              6.041667  
8560        63.32     96.436186      22.333333             28.059829  
8561         0.50     25.252525       1.000000              1.337838  
8562

In [9]:
import math

# Distribution plots for Numerical Columns
numerical_cols = df.select_dtypes(include=np.number).columns

ncols = 4
nrows = math.ceil(len(numerical_cols) / ncols)

fig = plt.figure(figsize=(15, 10))

for i, col in enumerate(numerical_cols):
    plt.subplot(nrows, ncols, i + 1)   # dynamic grid (no more 16-limit error)
    sns.histplot(df[col], kde=True, bins=30)
    plt.title(col)

plt.tight_layout()

# close the figure so it doesn't display and no unnecessary output remains
plt.close(fig)


In [10]:
os.makedirs("outputs/countplot", exist_ok=True)

categorical_cols = ["VendorName", "Description"]

plt.figure(figsize=(12,5))

for i, col in enumerate(categorical_cols):
    plt.subplot(1, 2, i+1)
    sns.countplot(
        y=df[col],
        order=df[col].value_counts().index[:10]
    )
    plt.title(f"Top 10 {col}")

plt.tight_layout()
plt.savefig("outputs/countplot/top_categorical_counts.png",
            dpi=300, bbox_inches="tight")
plt.close()

In [11]:
# Correlation Heatmap
# Create folder
save_dir = r"outputs\Correlation"
os.makedirs(save_dir, exist_ok=True)

# Correlation matrix
numerical_cols = df.select_dtypes(include=np.number).columns
correlation_matrix = df[numerical_cols].corr()

# Plot
plt.figure(figsize=(12,8))
sns.heatmap(
    correlation_matrix,
    annot=True,
    fmt=".2f",
    cmap="coolwarm",
    linewidths=0.5
)

plt.title("Correlation Heatmap")

# Save image
plt.savefig(
    os.path.join(save_dir, "correlation_heatmap.png"),
    dpi=300,
    bbox_inches="tight"
)

plt.close()   # closes figure (important)


## **Correlation Insights**
- PurchasePrice has weak correlations with TotalSalesDollars (-0.012) and GrossProfit (-0.016), suggesting that price variations do not significantly impact sales revenue or profit.
- Strong correlation between total purchase quantity and total sales quantity (0.999), confirming efficient inventory turnover.
- Negative correlation between profit margin & total sales price (-0.179) suggests that as sales price increases, margins decrease, possibly due to competitive pricing pressures.
- StockTurnover has weak negative correlations with both GrossProfit (-0.038) and ProfitMargin (-0.055), indicating that faster turnover does not necessarily result in higher profitability.

## **Data Analysis**
Identify Brands that needs Promotional or Pricing Adjustments which exhibit lower sales performance but higher profit margins.

In [12]:
brand_performance = df.groupby('Description').agg({
    'TotalSalesDollars': 'sum',
    'ProfitMargin':'mean'}).reset_index()

In [13]:
low_sales_threshold = brand_performance['TotalSalesDollars'].quantile(0.15)
high_margin_threshold = brand_performance['ProfitMargin'].quantile(0.85)

In [14]:
low_sales_threshold

np.float64(560.299)

In [15]:
high_margin_threshold

np.float64(64.97017552750113)

In [16]:
# Filter brands with low sales but high profit margins
target_brands = brand_performance[
    (brand_performance['TotalSalesDollars'] <= low_sales_threshold) &
    (brand_performance['ProfitMargin'] >= high_margin_threshold)
    ]
print("Brands with Low Sales but High Profit Margins:")
display(target_brands.sort_values('TotalSalesDollars'))

Brands with Low Sales but High Profit Margins:


Description  TotalSalesDollars  ProfitMargin
6199    Santa Rita Organic Svgn Bl               9.99     66.466466
2369             Debauchery Pnt Nr              11.58     65.975820
2070   Concannon Glen Ellen Wh Zin              15.95     83.448276
2188             Crown Royal Apple              27.86     89.806174
6237  Sauza Sprklg Wild Berry Marg              27.96     82.153076
...                            ...                ...           ...
5074   Nanbu Bijin Southern Beauty             535.68     76.747312
2271         Dad's Hat Rye Whiskey             538.89     81.851584
57       A Bichot Clos Marechaudes             539.94     67.740860
6245     Sbragia Home Ranch Merlot             549.75     66.444748
3326     Goulee Cos d'Estournel 10             558.87     69.434752

[198 rows x 3 columns]

In [17]:
from pathlib import Path
import os

# Base output folder (creates: <your-notebook-folder>/outputs )
BASE_OUT = Path.cwd() / "outputs"
BASE_OUT.mkdir(parents=True, exist_ok=True)

print("BASE_OUT set to:", BASE_OUT.resolve())

BASE_OUT set to: C:\Users\lenovo\outputs


In [18]:
save_dir = os.path.join(BASE_OUT, "Scatterplot")
os.makedirs(save_dir, exist_ok=True)

# 1) Build brand_performance (Brand-level)
brand_performance = (
    df.groupby("Brand", as_index=False)
      .agg(TotalSalesDollars=("TotalSalesDollars", "sum"),
           ProfitMargin=("ProfitMargin", "mean"))
)

# 2) Thresholds (use YOUR existing ones if already defined)
high_margin_threshold = 65   # change only if your notebook uses a different value
low_sales_threshold   = 1000 # change only if your notebook uses a different value

# 3) Target brands
target_brands = brand_performance[
    (brand_performance["TotalSalesDollars"] < low_sales_threshold) &
    (brand_performance["ProfitMargin"] > high_margin_threshold)
]

# 4) Tutor's visualization filter (IMPORTANT)
brand_performance = brand_performance[brand_performance["TotalSalesDollars"] < 10000]  # for better visualization

# 5) Plot EXACTLY like tutor
plt.figure(figsize=(10, 6))

sns.scatterplot(
    data=brand_performance,
    x="TotalSalesDollars",
    y="ProfitMargin",
    color="blue",
    label="All Brands",
    alpha=0.2
)

sns.scatterplot(
    data=target_brands,
    x="TotalSalesDollars",
    y="ProfitMargin",
    color="red",
    label="Target Brands"
)

plt.axhline(high_margin_threshold, linestyle="--", color="black", label="High Margin Threshold")
plt.axvline(low_sales_threshold, linestyle="--", color="black", label="Low Sales Threshold")

plt.xlabel("Total Sales ($)")
plt.ylabel("Profit Margin (%)")
plt.title("Brands for Promotional or Pricing Adjustments")
plt.legend()
plt.grid(True)

out_path = os.path.join(save_dir, "brands_promotional_pricing_scatter.png")
plt.savefig(out_path, dpi=300, bbox_inches="tight")
#plt.show()   # keep show if you want to see it in notebook
plt.close() 

 

### **which vendors and brands demonstrate the highest sales performance?**

In [19]:
def format_dollars(value):
    if value >= 1_000_000:
        return f"{value/1_000_000: .2f}M"
    elif value >= 1_000:
        return f"{value / 1_000: .2f}K"
    else:
        return str(value)

In [20]:
# Top Vendors and Brands by Sales Performance 
top_vendors = df.groupby("VendorName")["TotalSalesDollars"].sum().nlargest(10)
top_brands = df.groupby("Description")["TotalSalesDollars"].sum().nlargest(10)
top_vendors

VendorName
DIAGEO NORTH AMERICA INC      67990099.42
MARTIGNETTI COMPANIES         39330359.36
PERNOD RICARD USA             32063196.19
JIM BEAM BRANDS COMPANY       31423020.46
BACARDI USA INC               24854817.14
CONSTELLATION BRANDS INC      24218745.65
E & J GALLO WINERY            18399899.46
BROWN-FORMAN CORP             18247230.65
ULTRA BEVERAGE COMPANY LLP    16502544.31
M S WALKER INC                14706458.51
Name: TotalSalesDollars, dtype: float64

In [21]:
top_brands

Description
Jack Daniels No 7 Black    7964746.76
Tito's Handmade Vodka      7399657.58
Grey Goose Vodka           7209608.06
Capt Morgan Spiced Rum     6356320.62
Absolut 80 Proof           6244752.03
Jameson Irish Whiskey      5715759.69
Ketel One Vodka            5070083.56
Baileys Irish Cream        4150122.07
Kahlua                     3604858.66
Tanqueray                  3456697.90
Name: TotalSalesDollars, dtype: float64

In [22]:
top_brands.apply(lambda x : format_dollars(x))

Description
Jack Daniels No 7 Black     7.96M
Tito's Handmade Vodka       7.40M
Grey Goose Vodka            7.21M
Capt Morgan Spiced Rum      6.36M
Absolut 80 Proof            6.24M
Jameson Irish Whiskey       5.72M
Ketel One Vodka             5.07M
Baileys Irish Cream         4.15M
Kahlua                      3.60M
Tanqueray                   3.46M
Name: TotalSalesDollars, dtype: object

In [23]:
barplot_dir = os.path.join(BASE_OUT, "Barplot")
os.makedirs(barplot_dir, exist_ok=True)

print("Barplot folder:", barplot_dir)

Barplot folder: C:\Users\lenovo\outputs\Barplot


In [24]:
#below is the code used :  
plt.figure(figsize=(15, 5))


# -----------------------------
# Plot for Top Vendors
# -----------------------------
plt.subplot(1, 2, 1)
ax1 = sns.barplot(
    y=top_vendors.index,
    x=top_vendors.values,
    palette="Blues_r"
)
plt.title("Top 10 Vendors by Sales")

for bar in ax1.patches:
    ax1.text(
        bar.get_width() + (bar.get_width() * 0.02),
        bar.get_y() + bar.get_height() / 2,
        format_dollars(bar.get_width()),
        ha='left',
        va='center',
        fontsize=10,
        color='black'
    )

# -----------------------------
# Plot for Top Brands
# -----------------------------
plt.subplot(1, 2, 2)
ax2 = sns.barplot(
    y=top_brands.index.astype(str),
    x=top_brands.values,
    palette="Reds_r"
)
plt.title("Top 10 Brands by Sales")

for bar in ax2.patches:
    ax2.text(
        bar.get_width() + (bar.get_width() * 0.02),
        bar.get_y() + bar.get_height() / 2,
        format_dollars(bar.get_width()),
        ha='left',
        va='center',
        fontsize=10,
        color='black'
    )

plt.tight_layout()
out_path = os.path.join(barplot_dir, "top_vendors_top_brands.png")
plt.savefig(out_path, dpi=300, bbox_inches="tight")
plt.close()
plt.show()

### **Which vendors contribute the most to total purchase dollars?**

In [25]:
vendor_performance = df.groupby('VendorName').agg({
    'TotalPurchaseDollars':'sum',
    'GrossProfit':'sum',
    'TotalSalesDollars':'sum'
}).reset_index()
vendor_performance.shape

(119, 4)

In [26]:
vendor_performance['PurchaseContribution%'] = vendor_performance['TotalPurchaseDollars']/vendor_performance['TotalPurchaseDollars'].sum()*100

In [27]:
vendor_performance = round(vendor_performance.sort_values('PurchaseContribution%', ascending = False),2)

In [28]:
# Display Top 10 Vendors
top_vendors = vendor_performance.head(10)
top_vendors['TotalSalesDollars'] = top_vendors['TotalSalesDollars'].apply(format_dollars)
top_vendors['TotalPurchaseDollars'] = top_vendors['TotalPurchaseDollars'].apply(format_dollars)
top_vendors['GrossProfit'] = top_vendors['GrossProfit'].apply(format_dollars)
top_vendors

VendorName TotalPurchaseDollars GrossProfit  \
25     DIAGEO NORTH AMERICA INC               50.10M      17.89M   
57        MARTIGNETTI COMPANIES               25.50M      13.83M   
68            PERNOD RICARD USA               23.85M       8.21M   
46      JIM BEAM BRANDS COMPANY               23.49M       7.93M   
6               BACARDI USA INC               17.43M       7.42M   
20     CONSTELLATION BRANDS INC               15.27M       8.95M   
11            BROWN-FORMAN CORP               13.24M       5.01M   
30           E & J GALLO WINERY               12.07M       6.33M   
106  ULTRA BEVERAGE COMPANY LLP               11.17M       5.34M   
53               M S WALKER INC                9.76M       4.94M   

    TotalSalesDollars  PurchaseContribution%  
25             67.99M                  16.30  
57             39.33M                   8.30  
68             32.06M                   7.76  
46             31.42M                   7.64  
6              24.85M                   5.67  
20             24.22M                   4.97  
11             18.25M                   4.31  
30             18.40M                   3.93  
106            16.50M                   3.63  
53             14.71M                   3.18

In [29]:
top_vendors['Cumulative_Contribution%'] = top_vendors['PurchaseContribution%'].cumsum()
top_vendors

VendorName TotalPurchaseDollars GrossProfit  \
25     DIAGEO NORTH AMERICA INC               50.10M      17.89M   
57        MARTIGNETTI COMPANIES               25.50M      13.83M   
68            PERNOD RICARD USA               23.85M       8.21M   
46      JIM BEAM BRANDS COMPANY               23.49M       7.93M   
6               BACARDI USA INC               17.43M       7.42M   
20     CONSTELLATION BRANDS INC               15.27M       8.95M   
11            BROWN-FORMAN CORP               13.24M       5.01M   
30           E & J GALLO WINERY               12.07M       6.33M   
106  ULTRA BEVERAGE COMPANY LLP               11.17M       5.34M   
53               M S WALKER INC                9.76M       4.94M   

    TotalSalesDollars  PurchaseContribution%  Cumulative_Contribution%  
25             67.99M                  16.30                     16.30  
57             39.33M                   8.30                     24.60  
68             32.06M                   7.76                     32.36  
46             31.42M                   7.64                     40.00  
6              24.85M                   5.67                     45.67  
20             24.22M                   4.97                     50.64  
11             18.25M                   4.31                     54.95  
30             18.40M                   3.93                     58.88  
106            16.50M                   3.63                     62.51  
53             14.71M                   3.18                     65.69

In [30]:
 print(df.columns.tolist())


['VendorNumber', 'VendorName', 'Brand', 'Description', 'PurchasePrice', 'ActualPrice', 'Volume', 'TotalPurchaseQuantity', 'TotalPurchaseDollars', 'TotalSalesQuantity', 'TotalSalesDollars', 'TotalSalesPrice', 'TotalExciseTax', 'FreightCost', 'GrossProfit', 'ProfitMargin', 'StockTurnover', 'SalesToPurchaseRatio']


In [31]:
TOP_N = 10

top_vendors = (
    df.groupby("VendorName")["TotalPurchaseDollars"]
      .sum()
      .sort_values(ascending=False)
      .head(TOP_N)
)

# % contribution (of total purchases)
top_vendors = (top_vendors / df["TotalPurchaseDollars"].sum()) * 100

top_vendors = top_vendors.reset_index()
top_vendors.columns = ["VendorName", "PurchaseContribution%"]
top_vendors["Cumulative_Contribution%"] = top_vendors["PurchaseContribution%"].cumsum()

In [32]:
fig, ax1 = plt.subplots(figsize=(10, 6))

sns.barplot(
    x=top_vendors["VendorName"],
    y=top_vendors["PurchaseContribution%"],
    palette="mako",
    ax=ax1
)

# ✅ Put labels ABOVE bars (prevents overlap)
ymax = top_vendors["PurchaseContribution%"].max()
offset = ymax * 0.03  # 3% of max height

for i, v in enumerate(top_vendors["PurchaseContribution%"]):
    ax1.text(
        i,
        v + offset,
        f"{v:.2f}%",
        ha="center",
        va="bottom",
        fontsize=10,
        color="black"
    )

# Line plot
ax2 = ax1.twinx()
ax2.plot(
    top_vendors["VendorName"],
    top_vendors["Cumulative_Contribution%"],
    color="red",
    marker="o",
    linestyle="dashed",
    label="Cumulative Contribution %"
)

# Labels & title
ax1.set_xticklabels(top_vendors["VendorName"], rotation=90)
ax1.set_ylabel("Purchase Contribution %", color="blue")
ax2.set_ylabel("Cumulative Contribution %", color="red")
ax1.set_xlabel("Vendors")
ax1.set_title("Pareto Chart: Vendor Contribution to Total Purchases")

ax2.axhline(y=100, color="gray", linestyle="dashed", alpha=0.7)
ax2.legend(loc="upper right")

plt.tight_layout()

plt.savefig("outputs/pareto_charts/pareto_vendor_contribution.png", dpi=300, bbox_inches="tight")
 
plt.close(fig)


### **How much of total procurement is dependent on the top vendors?**

In [33]:
print(f"Total Purchase Contribution of Top 10 vendors is {round(top_vendors['PurchaseContribution%'].sum(),2)}%")

Total Purchase Contribution of Top 10 vendors is 65.69%


In [34]:
# Folder (change if your donut folder name is different)
# -------------------------------
donut_folder = "outputs/donutcharts"
os.makedirs(donut_folder, exist_ok=True)

# -------------------------------
# Prepare correct data from df
# -------------------------------
TOP_N = 10

vendor_totals = (
    df.groupby("VendorName")["TotalPurchaseDollars"]
      .sum()
      .sort_values(ascending=False)
)

top10_totals = vendor_totals.head(TOP_N)
total_purchase = vendor_totals.sum()

top10_pct = (top10_totals / total_purchase) * 100
other_pct = 100 - top10_pct.sum()

vendors = list(top10_pct.index)
purchase_contributions = list(top10_pct.values)

# Append "Other Vendors"
vendors.append("Other Vendors")
purchase_contributions.append(other_pct)

total_contribution = sum(purchase_contributions)  # will be ~100

# -------------------------------
# Donut Chart
# -------------------------------
fig, ax = plt.subplots(figsize=(8, 8))
wedges, texts, autotexts = ax.pie(
    purchase_contributions,
    labels=vendors,
    autopct='%1.1f%%',
    startangle=140,
    pctdistance=0.85,
    colors=plt.cm.Paired.colors
)

# Donut hole
centre_circle = plt.Circle((0, 0), 0.70, fc='white')
ax.add_artist(centre_circle)

# Center text
ax.text(
    0, 0,
    f"Top 10 Total:\n{(100 - other_pct):.2f}%",
    fontsize=14,
    fontweight='bold',
    ha='center',
    va='center'
)

ax.set_title("Top 10 Vendor's Purchase Contribution (%)")
plt.tight_layout()

# -------------------------------
# Save + Show + Close
# -------------------------------
save_path = os.path.join(donut_folder, "top10_vendor_purchase_contribution_donut.png")
plt.savefig(save_path, dpi=300, bbox_inches="tight")

plt.close(fig)

### **Does purchasing in bulk reduce the unit price, and what is the optimal purchase volume for cost savings?**

In [35]:
df['UnitPurchasePrice'] = df['TotalPurchaseDollars']/df['TotalPurchaseQuantity']

In [36]:
df["OrderSize"] = pd.qcut(df["TotalPurchaseQuantity"], q = 3, labels=["Small", "Medium", "Large"])

In [37]:
df[['OrderSize','TotalPurchaseQuantity']]

OrderSize  TotalPurchaseQuantity
0        Large                 145080
1        Large                 164038
2        Large                 187407
3        Large                 201682
4        Large                 138109
...        ...                    ...
8559     Small                      2
8560     Small                      6
8561     Small                      2
8562     Small                      1
8563     Small                      1

[8564 rows x 2 columns]

In [38]:
df.groupby('OrderSize')[['UnitPurchasePrice']].mean()

UnitPurchasePrice
OrderSize                   
Small              39.068186
Medium             15.486414
Large              10.777625

In [39]:
plt.figure(figsize=(10,6))
sns.boxplot(data=df, x="OrderSize", y="UnitPurchasePrice", palette="Set2")
plt.title("Impact of Bulk Purchasing on Unit Price")
plt.xlabel("Order Size")
plt.ylabel("Average Unit Purchase Price")

# ✅ save in boxplots folder
os.makedirs("outputs/boxplots", exist_ok=True)
plt.tight_layout()
plt.savefig("outputs/boxplots/impact_of_bulk_purchasing_on_unit_price.png", dpi=300, bbox_inches="tight")

plt.close()

- Vendors buying in bulk (Large Order Size) get the lowest unit price ($10.78 per unit), meaning higher margins if they can manage inventory efficiently.
- The price difference between Small and Large orders is substantial (~72% reduction in unit cost).
- This suggests that bulk pricing strategies successfully encourage vendors to purchase in larger volumes, leading to higher overall sales despite lower per-unit revenue.

### **Which vendors have low inventory turnover, indicating excess stock and slow-moving products?**

In [40]:
df[df['StockTurnover']<1].groupby('VendorName')[['StockTurnover']].mean().sort_values('StockTurnover', ascending = True).head(10)

StockTurnover
VendorName                                
ALISA CARR BEVERAGES              0.615385
HIGHLAND WINE MERCHANTS LLC       0.708333
PARK STREET IMPORTS LLC           0.751306
Circa Wines                       0.755676
Dunn Wine Brokers                 0.766022
CENTEUR IMPORTS LLC               0.773953
SMOKY QUARTZ DISTILLERY LLC       0.783835
TAMWORTH DISTILLING               0.797078
THE IMPORTED GRAPE LLC            0.807569
WALPOLE MTN VIEW WINERY           0.820548

### **How much capital is locked in unsold inventory per vendor, and which vendors contribute the most to it?**

In [41]:
df["UnsoldInventoryValue"] = (df["TotalPurchaseQuantity"] - df["TotalSalesQuantity"]) * df["PurchasePrice"]
print('Total Unsold Capital:' , format_dollars(df["UnsoldInventoryValue"].sum()))

Total Unsold Capital:  2.71M


In [42]:
# Aggregate Capital Locked per Vendor
inventory_value_per_vendor = df.groupby("VendorName")["UnsoldInventoryValue"].sum().reset_index()

inventory_value_per_vendor = inventory_value_per_vendor.sort_values(by="UnsoldInventoryValue", ascending=False)
inventory_value_per_vendor["UnsoldInventoryValue"] = inventory_value_per_vendor["UnsoldInventoryValue"].apply(format_dollars)
inventory_value_per_vendor.head(10)

VendorName UnsoldInventoryValue
25   DIAGEO NORTH AMERICA INC              722.21K
46    JIM BEAM BRANDS COMPANY              554.67K
68          PERNOD RICARD USA              470.63K
116  WILLIAM GRANT & SONS INC              401.96K
30         E & J GALLO WINERY              228.28K
79             SAZERAC CO INC              198.44K
11          BROWN-FORMAN CORP              177.73K
20   CONSTELLATION BRANDS INC              133.62K
61      MOET HENNESSY USA INC              126.48K
77     REMY COINTREAU USA INC              118.60K

### **What is the 95% confidence intervals for profit margins of top-performing and low-performing vendors.**

In [43]:
# Determine thresholds for top and low performing vendors
top_threshold = df["TotalSalesDollars"].quantile(0.75)
low_threshold = df["TotalSalesDollars"].quantile(0.25)

In [44]:
# Filter top and low vendors and extract ProfitMargin values
top_vendors = df[df["TotalSalesDollars"] >= top_threshold]["ProfitMargin"].dropna()
low_vendors = df[df["TotalSalesDollars"] <= low_threshold]["ProfitMargin"].dropna()

In [45]:
top_vendors

0       25.297693
1       21.062810
2       24.675786
3       27.139908
4       28.412764
          ...    
3523    79.684817
3681    85.782102
4751    93.085860
4920    95.012530
5050    94.271857
Name: ProfitMargin, Length: 2141, dtype: float64

In [46]:
low_vendors

5631     4.111764
5652     6.145626
5701    12.007271
5704     1.677308
5724     7.239599
          ...    
8559    83.448276
8560    96.436186
8561    25.252525
8562    98.974037
8563    99.166079
Name: ProfitMargin, Length: 2141, dtype: float64

In [47]:
# ================================
# Confidence Interval Function
# ================================
def confidence_interval(data, confidence=0.95):
    mean_val = np.mean(data)
    std_err = np.std(data, ddof=1) / np.sqrt(len(data))  # Standard error
    t_critical = stats.t.ppf((1 + confidence) / 2, df=len(data) - 1)
    margin_of_error = t_critical * std_err
    return mean_val, mean_val - margin_of_error, mean_val + margin_of_error


# ================================
# Calculate Confidence Intervals
# ================================
top_mean, top_lower, top_upper = confidence_interval(top_vendors)
low_mean, low_lower, low_upper = confidence_interval(low_vendors)

print(f"Top Vendors 95% CI: ({top_lower:.2f}, {top_upper:.2f}), Mean: {top_mean:.2f}")
print(f"Low Vendors 95% CI: ({low_lower:.2f}, {low_upper:.2f}), Mean: {low_mean:.2f}")


# ================================
# Plot
# ================================
plt.figure(figsize=(12, 6))

# ---- Top Vendors Plot ----
sns.histplot(
    top_vendors,
    kde=True,
    color="blue",
    bins=30,
    alpha=0.5,
    label="Top Vendors"
)
plt.axvline(top_lower, color="blue", linestyle="--", label=f"Top Lower: {top_lower:.2f}")
plt.axvline(top_upper, color="blue", linestyle="--", label=f"Top Upper: {top_upper:.2f}")
plt.axvline(top_mean, color="blue", linestyle="-", label=f"Top Mean: {top_mean:.2f}")

# ---- Low Vendors Plot ----
sns.histplot(
    low_vendors,
    kde=True,
    color="red",
    bins=30,
    alpha=0.5,
    label="Low Vendors"
)
plt.axvline(low_lower, color="red", linestyle="--", label=f"Low Lower: {low_lower:.2f}")
plt.axvline(low_upper, color="red", linestyle="--", label=f"Low Upper: {low_upper:.2f}")
plt.axvline(low_mean, color="red", linestyle="-", label=f"Low Mean: {low_mean:.2f}")

# ---- Finalize Plot ----
plt.title("Confidence Interval Comparison: Top vs. Low Vendors (Profit Margin)")
plt.xlabel("Profit Margin (%)")
plt.ylabel("Frequency")
plt.legend()
plt.grid(True)

# ================================
# Save + Show + Close (same pattern)
# ================================
os.makedirs("outputs/histplots", exist_ok=True)
plt.savefig(
    "outputs/histplots/ci_top_vs_low_vendors_profit_margin.png",
    dpi=300,
    bbox_inches="tight"
)
 
plt.close()


Top Vendors 95% CI: (30.74, 31.61), Mean: 31.18
Low Vendors 95% CI: (40.50, 42.64), Mean: 41.57


- The confidence interval for low-performing vendors (40.48% to 42.62%) is significantly higher than that of top-performing vendors (30.74% to 31.61%).
- This suggests that vendors with lower sales tend to maintain higher profit margins, potentially due to premium pricing or lower operational costs.
- For High-Performing Vendors: If they aim to improve profitability, they could explore selective price adjustments, cost optimization, or bundling strategies.
- For Low-Performing Vendors: Despite higher margins, their low sales volume might indicate a need for better marketing, competitive pricing, or improved distribution strategies.


### Is there a significant difference in profit margins between top-performing and low-performing vendors?

**Hypothesis:**

**H₀ (Null Hypothesis):** There is no significant difference in the mean profit margins of top-performing and low-performing vendors.

**H₁ (Alternative Hypothesis):** The mean profit margins of top-performing and low-performing vendors are significantly different.

In [48]:
top_threshold = df["TotalSalesDollars"].quantile(0.75)
low_threshold = df["TotalSalesDollars"].quantile(0.25)

top_vendors = df[df["TotalSalesDollars"] >= top_threshold]["ProfitMargin"].dropna()
low_vendors = df[df["TotalSalesDollars"] <= low_threshold]["ProfitMargin"].dropna()

# Perform Two-Sample T-Test
t_stat, p_value = ttest_ind(top_vendors, low_vendors, equal_var=False)

# Print results
print(f"T-Statistic: {t_stat:.4f}, P-Value: {p_value:.4f}")
if p_value < 0.05:
    print("Reject Ho: There is a significant difference in profit margins between top and low-performing vendors.")
else:
    print("Fail to Reject Ho: No significant difference in profit margins.")

T-Statistic: -17.6695, P-Value: 0.0000
Reject Ho: There is a significant difference in profit margins between top and low-performing vendors.


In [49]:
import sqlite3
import pandas as pd

# 1. Connect to your SQLite database
conn = sqlite3.connect("inventory.db")

# 2. Read the cleaned version of vendor_sales_summary
df_clean = pd.read_sql_query("""
    SELECT * 
    FROM vendor_sales_summary
    WHERE GrossProfit > 0
      AND ProfitMargin > 0
      AND TotalSalesQuantity > 0
""", conn)

# 3. Export to CSV for Power BI
df_clean.to_csv("vendor_sales_summary_clean.csv", index=False)

conn.close()
print("Exported cleaned data to vendor_sales_summary_clean.csv")

Exported cleaned data to vendor_sales_summary_clean.csv
